File: spl-net/src/main/java/bgu/spl171/net/impl/TFTP/CommandEncoders/DataCommandEncoder.java
package bgu.spl171.net.impl.TFTP.CommandEncoders;

import bgu.spl171.net.impl.TFTP.Commands.Command;
import bgu.spl171.net.impl.TFTP.Commands.Data;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.ShortEncoderDecoder;

public class DataCommandEncoder implements CommandEncoder {

    private ShortEncoderDecoder shortEncoderDecoder = new ShortEncoderDecoder();
    @Override
    public byte[] encode(Command cmd) {
        Data data = (Data)cmd;
        byte[] opCode = shortEncoderDecoder.encode((short)3);
        byte[] size = shortEncoderDecoder.encode(data.getSize());
        byte[] blockId = shortEncoderDecoder.encode(data.getBlockId());
        return EncoderUtils.mergeByteArray(opCode,size,blockId,data.getData(),new byte[]{0});
    }
}

File: java to c++ encoders/DataCommandEncoder.cpp
#include "DataCommandEncoder.h"
#include "EncoderUtils.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace CommandEncoders
					{
						using bgu::spl171::net::impl::TFTP::Commands::Command;
						using bgu::spl171::net::impl::TFTP::Commands::Data;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::ShortEncoderDecoder;

						std::vector<char> DataCommandEncoder::encode(Command *cmd)
						{
							Data *data = static_cast<Data*>(cmd);
							std::vector<char> opCode = shortEncoderDecoder->encode(data->getOpCode());
							std::vector<char> size = shortEncoderDecoder->encode(data->getSize());
							std::vector<char> blockId = shortEncoderDecoder->encode(data->getBlockId());
							return EncoderUtils::mergeByteArray(
							{
								opCode, size, blockId, data->getData(), std::vector<char>{0}
							});
						}
					}
				}
			}
		}
	}
}
